src/visitors/sympy_solver_visitor.cpp

```diff
@@ -186,26 +208,28 @@ void SympySolverVisitor::construct_eigen_solver_block(
 
     if (linear) {
         /// create eigen linear solver block
-        setup_x_eqs.insert(setup_x_eqs.end(), solutions.begin(), solutions.end());
-        auto setup_x_block = create_statement_block(setup_x_eqs);
+        setup_x_operator_eqs.insert(setup_x_operator_eqs.end(), solutions.begin(), solutions.end());
+        auto setup_x_operator_block = create_statement_block(setup_x_operator_eqs);
         auto solver_block = std::make_shared<ast::EigenLinearSolverBlock>(n_state_vars,
                                                                           variable_block,
                                                                           initialize_block,
-                                                                          setup_x_block,
+                                                                          setup_x_operator_block,
                                                                           update_state_block,
                                                                           finalize_block);
         /// replace statement block with solver block as it contains all statements
         ast::StatementVector solver_block_statements{
             std::make_shared<ast::ExpressionStatement>(solver_block)};
         block_with_expression_statements->set_statements(std::move(solver_block_statements));
     } else {
+        /// filter solutions for variable named "X" and change it to "X_operator"
+        auto solutions_filtered = filter_X(solutions);
```

Comment:
somehow this also doesn't do what the issue proposes, no? You always replace X by X_operator, rather than finding an alternative only when X is taken...